### ISSUE TITLE
custom_vjp (and jvp) doesn't support collectives

### ISSUE THREAD

**Issue by apaszke** [google/jax#7047]
The following crashes (reported by @sharadmv):
```py
@jax.custom_vjp
def f(x):
  return lax.psum(x, 'foo')

def f_fwd(x):
  return lax.psum(x, 'foo'), None

def f_bwd(res, dx):
  return dx
f.defvjp(f_fwd, f_bwd)

maps.xmap(f, axis_sizes={'foo': 4}, in_axes={}, out_axes={})(2.)
```

Most likely this is because the `custom_vjp` gets no batching tracers as inputs, entirely skipping the batching interpreter, while it really needs to get triggered to desugar the psum over the axis name into a positional one.
